### PR TITLE
[Notification] Notification and subscription

### PR DESCRIPTION
**Pull request by qnkhuat**
Epic: https://github.com/metabase/metabase/issues/47438

This resolves the first 2 points of the milestone 1:
- It adds 2 tables: notification and notification_subscription
- Introduces a new event type: `:event/notification` and makes sure all published events will trigger `event/notification`. This then checks for subscribed notifications and sends them.
- introduce a new module: notification. However, the functionality is currently limited.

In short, this PR tries to make the bridge between system event <-> notification.

In the next PR, we will then work on the notification <-> channel part.

### CODE REVIEW

File: src/metabase/cmd/copy.clj [hunk 1]
```diff
@@ -97,7 +97,10 @@
     :model/TablePrivileges
     :model/AuditLog
     :model/RecentViews
-    :model/UserParameterValue]
+    :model/UserParameterValue
+    ;; 49+
+    :model/Notification
```
Comment (qnkhuat) [inline src/metabase/cmd/copy.clj:102]: I'm not sure if we want to copy it? putting it here bc we copied Pulse. but I made a note in the epic to review this later.

File: resources/migrations/001_update_migrations.yaml [hunk 2]
```diff
@@ -9270,6 +9270,119 @@ databaseChangeLog:
...
+      author: qnkhuat
+      comment: Create the notification table
+      preConditions:
+        - not:
+          - tableExists:
+              tableName: notification
+      changes:
+        - createTable:
+            tableName: notification
+            remarks: join table that connect notification subscriptions and notification destinations
```
Comment (crisptrutski) [inline resources/migrations/001_update_migrations.yaml:9284]: Tracking our rename, also there's no need to repeat the qualifier.

```suggestion
            remarks: join table that connect notification subscriptions and handlers
```

File: src/metabase/models/interface.clj [hunk 3]
```diff
@@ -251,6 +251,28 @@
   {:in  json-in
    :out json-out-without-keywordization})
 
+(defn transform-validator
+  "Given a transform, returns a transform that call `assert-fn` on the value.
```
Comment (crisptrutski) [inline src/metabase/models/interface.clj:255]:
```suggestion
  "Given a transform, returns a transform that call `assert-fn` on the "out" value.
```

Not sure how better to refer to which side we validate on, but this helper is too ambiguous without qualifying it some way. 

I thought about perhaps "transformed value", but  the designation of `in` and `out` in our transforms feel so flipped to me (e.g. "out" being what we do to values coming *into* the application). It feels like I'm missing some analogy or other use case 🤔

File: src/metabase/models/notification.clj [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+(def ^:private subscription-types #{:notification-subscription/system-event})
+
+(t2/deftransforms :model/Notification
+  {:payload_type (mi/transform-validator mi/transform-keyword (partial assert-enum notification-types))})
+
+(t2/deftransforms :model/NotificationSubscription
+  {:type       (mi/transform-validator mi/transform-keyword (partial assert-enum subscription-types))
+   :event_name (mi/transform-validator mi/transform-keyword (fn [value]
+                                                              (when-not (= "event" (-> value keyword namespace))
+                                                                (throw (ex-info "Event name must be a namespaced keyword under :event" {:status-code 400
```
Comment (crisptrutski) [inline src/metabase/models/notification.clj:34]: nit: Line is super long - I think punt the `ex-data` to start on the next line.

File: src/metabase/models/notification.clj [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+(doseq [model [:model/Notification
+               :model/NotificationSubscription]]
+  (doto model
+    (derive :metabase/model)
+    (derive (if (= model :model/NotificationSubscription)
+              :hook/created-at-timestamped?
+              :hook/timestamped?))))
+
+(defn notifications-for-event
+  "Find all notifications for a given event."
```
Comment (crisptrutski) [inline src/metabase/models/notification.clj:46]:
```suggestion
  "Find all active notifications for a given event."
```

Perhaps even "for a given *system* event", given the additional filter. I'm not sure if we might widen the scope of this function if we add additional types.

File: src/metabase/models/notification.clj [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+                          [:= :ns.event_name (u/qualified-name event-name)]
+                          [:= :ns.type (u/qualified-name :notification-subscription/system-event)]]}))
+
+(defn create-notification!
+  "Create a new notification with `subsciptions`.
+  Return the created notification."
+  [notification subcriptions]
+  (t2/with-transaction [_conn]
+    (let [noti    (t2/insert-returning-instance! :model/Notification notification)
+          noti-id (:id noti)]
```
Comment (crisptrutski) [inline src/metabase/models/notification.clj:63]: nit: This abbreviation feels weird to me, I'd prefer `n` or `instance`, or just `notification`. Or avoiding any name at all, by using `prog1`.

I think just `id` is fine for the other binding, its meaning is clear in the context.

File: test/metabase/events/notification_test.clj [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+       ~@body
+       (finally
+         (doseq [topic# topics#]
+           (underive topic# :metabase/event))))))
+
+(deftest supported-events-with-notification-will-be-sent-test
+  (mt/with-model-cleanup [:model/Notification]
+    (with-temporary-event-topics #{:event/test-notification}
+      (let [topic      :event/test-notification
+            noti-1     (models.notification/create-notification!
```
Comment (crisptrutski) [inline test/metabase/events/notification_test.clj:28]: Again, I don't think this is a good abbreviation to introduce. `n1` seems fine in the context, and just `sent` for the accumulator.

File: test/metabase/events/notification_test.clj [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+                        {:payload_type :notification/system-event
+                         :active       true}
+                        [{:type       :notification-subscription/system-event
+                          :event_name topic}])
+            _inactive  (models.notification/create-notification!
+                        {:payload_type :notification/system-event
+                         :active       false}
+                        [{:type       :notification-subscription/system-event
+                          :event_name topic}])
+            sent-notis (atom #{})]
```
Comment (crisptrutski) [inline test/metabase/events/notification_test.clj:43]: I think its worth testing that we're not sending duplicates, i.e. make this hold an array, and use `sort` to make the assertion order independent.

File: test/metabase/events/notification_test.clj [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+        (testing "publishing event will send all the actively subscribed notifciations"
+          (with-redefs [notification/send-notification!      (fn [notification] (swap! sent-notis conj notification))
```
Comment (crisptrutski) [inline test/metabase/events/notification_test.clj:45]: Let's make this threadsafe using `mt/with-dynamic-redefs`.

File: test/metabase/models/notification_test.clj [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+   [metabase.test :as mt]
+   [toucan2.core :as t2]))
+
+;; ------------------------------------------------------------------------------------------------;;
+;;                                      Life cycle test                                            ;;
+;; ------------------------------------------------------------------------------------------------;;
+
+(deftest notification-type-test
+  (mt/with-model-cleanup [:model/Notification]
+    (testing "success path"
```
Comment (crisptrutski) [inline test/metabase/models/notification_test.clj:14]: Maybe use a more descriptive label?
```suggestion
    (testing "success if :payload_type is supported"
```

**Review by crisptrutski — APPROVED** [review]
Looks sane, and like all the validation and general structure. Just left some notes around clarity and slightly stronger testing.

I'm still somewhat uncomfortable with the name `notification`, as its very suggestive of messaging users, and my understanding is that we'll widen the scope of how these are used. It seems I'm the minority, and might come around 😄
